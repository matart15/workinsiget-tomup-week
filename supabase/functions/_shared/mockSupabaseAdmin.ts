const SUPABASE_FUNCTION_NAMES = [
  "from",
  "select",
  "eq",
  "delete",
  "update",
  "insert",
  "limit",
  "single",
] as const;
type ReturnFunctionName = (typeof SUPABASE_FUNCTION_NAMES)[number];

export const mockSupabaseAdmin: any = (
  returnValues: Record<ReturnFunctionName, any>[],
) => {
  let allCalls: any = [];
  let allArgs: any = [];
  let currentChainIndex = 0;

  const supabaseAdmin: any = {};
  for (const functionName of SUPABASE_FUNCTION_NAMES) {
    supabaseAdmin[functionName] = (...args: any[]) => {
      const returnValue = returnValues[currentChainIndex];
      const returnFunctionName: ReturnFunctionName = Object.keys(
        returnValue,
      )[0] as any;

      allArgs.push({ [functionName]: args });
      if (returnFunctionName === functionName) {
        allCalls.push(allArgs);
        currentChainIndex++;
        return returnValue[returnFunctionName];
      }
      return supabaseAdmin;
    };
  }

  return {
    allCalls,
    supabaseAdmin,
  };
};
