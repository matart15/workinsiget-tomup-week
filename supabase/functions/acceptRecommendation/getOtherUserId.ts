import { SupabaseClient } from "https://esm.sh/@supabase/supabase-js@2";

import { Database } from "../_shared/database.ts";

export const getOtherUserId = async ({
  recommendation_id,
  supabaseAdmin,
  currentUserId,
}: {
  recommendation_id: number;
  supabaseAdmin: SupabaseClient<Database>;
  currentUserId: string;
}) => {
  // check recommendation exists on recommendation table
  const recommendation = await supabaseAdmin
    .from("u_recommendations")
    .select("*")
    .eq("id", recommendation_id)
    .single();
  if (recommendation.error) {
    throw recommendation.error;
  }

  if (recommendation.data.user1_id === currentUserId) {
    return recommendation.data.user2_id;
  }
  return recommendation.data.user1_id;
};
