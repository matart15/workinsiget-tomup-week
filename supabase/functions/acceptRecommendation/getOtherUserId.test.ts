import {
  assertEquals,
  assertRejects,
} from "https://deno.land/std@0.206.0/testing/asserts.ts";
import { assertSnapshot } from "https://deno.land/std@0.206.0/testing/snapshot.ts";

import { mockSupabaseAdmin } from "../_shared/mockSupabaseAdmin.ts";
import { getOtherUserId } from "./getOtherUserId.ts";

Deno.test("getOtherUserId should return other user id", async (t) => {
  const recommendation_id = 123;
  const { supabaseAdmin, allCalls } = mockSupabaseAdmin([
    {
      single: {
        data: { user2_id: "456" },
      },
    },
  ]);

  // Act
  const result = await getOtherUserId({
    recommendation_id,
    supabaseAdmin,
  });

  // Assert
  assertEquals(result, "456");
  // assert all args
  await assertSnapshot(t, allCalls);
});

Deno.test(
  "getOtherUserId should throw an error if recommendation does not exist",
  async () => {
    const recommendation_id = 123;
    const { supabaseAdmin } = mockSupabaseAdmin([
      {
        single: {
          error: new Error("Recommendation not found"),
        },
      },
    ]);

    // Act and Assert
    await assertRejects(
      async () => {
        await getOtherUserId({
          recommendation_id,
          supabaseAdmin,
        });
      },
      Error,
      "Recommendation not found",
    );
  },
);
