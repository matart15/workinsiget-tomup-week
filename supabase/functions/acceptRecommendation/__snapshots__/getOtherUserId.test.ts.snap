export const snapshot = {};

snapshot[`getOtherUserId should return other user id 1`] = `
[
  [
    {
      from: [
        "u_recommendations",
      ],
    },
    {
      select: [
        "*",
      ],
    },
    {
      eq: [
        "id",
        123,
      ],
    },
    {
      single: [],
    },
  ],
]
`;
